src/main/java/io/aiven/kafka/connect/transforms/ConcatFields.java

```diff
@@ -0,0 +1,180 @@
+        }
+
+        final Optional<Object> newValue;
+        if (config.fieldName().isPresent()) {
+            newValue = getNewValueForNamedField(
+                    record.toString(), schemaAndValue.schema(), schemaAndValue.value(), config.fieldName().get());
+        } else {
+            newValue = getNewValueWithoutFieldName(
+                    record.toString(), schemaAndValue.schema(), schemaAndValue.value());
+        }
+
+        if (newValue.isPresent()) {
+            return record.newRecord(
+                    record.topic(),
+                    record.kafkaPartition(),
+                    record.keySchema(),
+                    record.key(),
+                    record.valueSchema(),
+                    newValue.get(),
+                    record.timestamp(),
+                    record.headers()
+            );
+        } else {
+            return record;
+        }
+ */
+    @Override
+    public R apply(final R record) {
+        final SchemaAndValue schemaAndValue = getSchemaAndValue(record);
+
+        if (schemaAndValue.value() == null) {
+            throw new DataException(dataPlace() + " Value can't be null: " + record);
+        }
+
+        final R newRecord;
+
+        if (schemaAndValue.value() instanceof Struct) {
+            final Struct struct = (Struct) schemaAndValue.value();
+            final SchemaBuilder newSchema = SchemaBuilder.struct();
+            final StringJoiner outputValue = new StringJoiner(config.delimiter());
+
+            if (schemaAndValue.schema() != null) {
+                schemaAndValue.schema().fields().forEach(field -> newSchema.field(field.name(), field.schema()));
+            } else {
+                struct.schema().fields().forEach(field -> newSchema.field(field.name(), field.schema()));
+            }
+            newSchema.field(config.outputFieldName(), Schema.OPTIONAL_STRING_SCHEMA);
+            final Struct newStruct = new Struct(newSchema);
+            struct.schema().fields().forEach(field -> {
+                newStruct.put(field.name(), struct.get(field));
+            });
+            config.fieldNames().forEach(field -> {
+                try {
+                    if (struct.get(field) == null) {
+                        outputValue.add(config.fieldReplaceMissing());
+                    } else {
+                        outputValue.add(struct.get(field).toString());
+                    }
+                } catch (final DataException e) {
+                    log.debug("{} is missing, concat will use {}", field, config.fieldReplaceMissing());
+                    outputValue.add(config.fieldReplaceMissing());
+                }
+            });
+            newStruct.put(config.outputFieldName(), outputValue.toString());
+            newRecord = createNewRecord(record, newSchema.build(), newStruct);
+        } else if (schemaAndValue.value() instanceof Map) {
+            final Map newValue = (Map<?, ?>) schemaAndValue.value();
```

Comment:
This mutates the old values in place, which combined with other issues make the tests green, while the code is incorrect. Please make a copy of it like:
```suggestion
            final Map<?, ?> newValue = new HashMap<>((Map<?, ?>) schemaAndValue.value());
```